StylePropertyMap: Updated the get method to work with properties that contain a sequence of StyleValues

njlawton opened the pull request.

Commit:
Implemented the function isSupportedStyleValue in the PropertyDictionary class
Tests were also made to validate its input

Commit:
made outline for helper function used to parse sequences of style values

Commit:
Merge branch 'keyword-parse' into Style-map-get-sequence

Commit:
Implemented Sequence support for the get method
This will only work if the sequence is separated by ', ' other wise will throw an error

Commit:
Fixed error in processing KeywordValues in the get method

Commit:
Removed a unneeded console.log from the StyleValue class
Also fixed small code formmating issues

Commit:
Merge branch 'master' into Style-map-get-sequence

Commit:
removed isSupportedStyleValue method from the PropertyDictionary class
This method has gone unused in the changes made for this commit and is therefore irrelevant to this commit

Commit:
Readded code mistakenly lost in merging branches

wilddamon commented on src/style-property-map-readonly.js:
```diff
@@ -29,7 +29,7 @@
 
     var parsed = StyleValue.parse(property, propertyString);
-    if (typeof parsed == 'Array') {
+    if (parsed.constructor == Array) {
```
Out of curiosity, why this change?

njlawton replied:
When looking through JavaScript resources this was suggested as a more common way of doing this. This is no big deal either way and I will change it if you would like me to.

wilddamon replied:
Best to keep it consistent with everything else we have.

njlawton replied:
As it turns out you cant use typeof to refer to arrays as the function will return 'object'

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
-      throw new TypeError('Property name must be a string');
-    }
-    if (typeof cssString != 'string') {
-      throw new TypeError('Must parse a string');
```
You seem to have lost some of my changes.

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
+    for (var i = 0; i < valueArray.length; i++) {
+      var cssString = valueArray[i];
+      cssString = cssString.trim().toLowerCase();
+      if (typeof cssString != 'string') {    
```
The check that value is a string must be before you call split - otherwise you will get an exception on line 29 before you get to this line.

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
   function StyleValue() {}
 
-  // TODO: Add support for cssString being a sequence.
-  StyleValue.parse = function(property, cssString) {
```
I preferred calling this cssString rather than value here. Please put that back :)

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
-    if (typeof property != 'string') {
-      throw new TypeError('Property name must be a string');
-    }
-    if (typeof cssString != 'string') {
```
You need to keep this check here.

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
+        }
+      }
+
+      if (successfulParse == false) {
```
``` javascript
if (!successfulParse) {
```

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
-      if (styleValueObject != null) {
-        return styleValueObject;
+      var styleValueObject = null;
+      var successfulParse = false;
```
You can define this outside the loop to avoid initialising it many times.

wilddamon commented on src/style-value.js:
```diff
@@ -16,34 +16,55 @@
+  };
+
+  StyleValue.prototype.
+      _getValueArray = function(property, value) {
```
This function doesn't seem to be used?

wilddamon commented on test/js/computed-style-property-map.js:
```diff
@@ -23,6 +23,18 @@ suite('Computed StylePropertyMap', function() {
+
+    assert.instanceOf(propertyStyleValue, NumberValue);
+    assert.strictEqual(propertyStyleValue.cssString, '4');
+  });
```
It would also be great to have a test for the parse method directly - you can create a new test file for StyleValue.

njlawton replied:
Done

Commit:
Fixed problems in the StyleValue class found in the pull request reveiw

Commit:
added a try catch statement to StyleValues parse method
This is to ensure that parse methods that throw errors will not end the parse function

Commit:
Added tests for the StyleValue parse method

Commit:
Removed method in StyleValue that has gone unused due to refactoring

Commit:
Fixed small formatting issues in the StyleValue class

wilddamon commented on src/style-value.js:
```diff
@@ -29,21 +28,44 @@
-    if (propertyDictionary().isValidKeyword(property, cssString)) {
-      return new KeywordValue(cssString);
-    }
+    //Currently only supports sequences separated by ', '
```
Nitpick: needs a space after //

wilddamon commented on src/style-value.js:
```diff
@@ -29,21 +28,44 @@
+    var successfulParse = false;
+    for (var i = 0; i < valueArray.length; i++) {
+      var cssStringStyleValue = valueArray[i];
+      cssStringStyleValue = cssStringStyleValue.trim().toLowerCase();
```
I think it'd be better to do the toLowerCase() on the whole cssString before you split it. Then you only need to trim here.

wilddamon commented on src/style-value.js:
```diff
@@ -29,21 +28,44 @@
+        try {
+          styleValueObject = supportedStyleValues[j].parse(cssStringStyleValue);
+        } catch (e) {
+          //Ensures method does not terminate if a StyleValue parse method throws an error
```
Same nitpick as earlier in the file. Needs space after //

wilddamon commented:
LGTM once nitpicks are fixed.

Commit:
Fixed small problems found in code reveiw
> Added test case to test that the parse method is case insensitive

wilddamon commented:
LGTM
